src/instructlab/model/backends/vllm.py

```diff
@@ -0,0 +1,194 @@
+                self.model_path.as_posix(),
+                self.model_family,
+                self.host,
+                self.port,
+            )
+        except ServerException as exc:
+            raise exc
+
+    def create_server_process(self, port: str) -> multiprocessing.Process:
+        mpctx = multiprocessing.get_context(None)
+        server_process = mpctx.Process(
+            target=server,
+            kwargs={
+                "model_path": self.model_path.as_posix(),
+                "model_family": self.model_family,
+                "port": port,
+                "host": self.host,
+            },
+        )
+
+        return server_process
+
+    def run_detached(
+        self, tls_insecure, tls_client_cert, tls_client_key, tls_client_passwd
+    ):
+        try:
+            server_process, api_base = ensure_server(
+                logger=self.logger,
+                server_process_func=self.create_server_process,
+                api_base=self.api_base,
+                host=self.host,
+                port=self.port,
+                tls_insecure=tls_insecure,
+                tls_client_cert=tls_client_cert,
+                tls_client_key=tls_client_key,
+                tls_client_passwd=tls_client_passwd,
+            )
+            self.process = server_process
+            self.api_base = api_base
+        except ServerException as exc:
+            raise exc
+
+    def shutdown(self):
+        """Shutdown vllm server"""
+        # Needed when the chat is connected to a server and the temporary server is started
+        if self.process is not None:
+            self.logger.debug("Terminating temporary server.")
+            self.process.terminate()
+            self.process.join(timeout=30)
+
+        else:
+            self.logger.debug("No server process to terminate.")
+
+
+# Heavily inspired by https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/api_server.py
+def server(
+    model_path,
+    model_family,
+    host="localhost",
+    port=8000,
+):
+    # Third Party
+    from vllm.engine.arg_utils import AsyncEngineArgs
+    from vllm.engine.async_llm_engine import AsyncLLMEngine
+    from vllm.entrypoints.openai.serving_chat import OpenAIServingChat
+    from vllm.entrypoints.openai.serving_completion import OpenAIServingCompletion
+    from vllm.usage.usage_lib import UsageContext
+    import vllm.entrypoints.openai.api_server as api_server
+
+    app = fastapi.FastAPI()
+
+    engine_args = AsyncEngineArgs(
+        model=model_path,
+        max_model_len=4096,  # TODO: make this configurable
+    )
+
+    engine = api_server.engine = AsyncLLMEngine.from_engine_args(
+        engine_args, usage_context=UsageContext.OPENAI_API_SERVER
+    )
+    engine.engine_use_ray = False
+
+    model_config = asyncio.run(engine.get_model_config())
+
+    template = ""
+    # TODO: find the equivalent in vllm for eos_token and bos_token
+    eos_token = "<|endoftext|>"
+    bos_token = ""
```

Comment:
@leseb are `eos_token` and `bos_token` just here for the TODO? They don't seem to lead anywhere

Reply:
Yes, just TODO.